### ISSUE TITLE
Problem with Common.Logging

### ISSUE THREAD

**Issue by dtosato** [NightOwl888/BoboBrowse.Net#11]
@NightOwl888 
Hi, I found that the `Common.Logging >= 1.2` dependence written into the NuGet package of Bobo is not valid. In fact I use   `Common.Logging 3.0.0` in my solution and it does not work correctly.
It returns the following runtime error:
```
System.MissingMethodException
HResult=-2146233069
Message=Method not found: 'Common.Logging.ILog Common.Logging.LogManager.GetLogger(System.Type)'
Source=BoboBrowse.Net
StackTrace: in BoboBrowse.Net.Facets.Impl.SimpleFacetHandler..cctor()
InnerException: 
```

**Comment by NightOwl888** [comment]
Thanks for the report.

Upon researching this issue, I discovered that this is a [known problem with NuGet](http://nuget.codeplex.com/discussions/578411). More information [here](https://github.com/spring-projects/spring-net/issues/92).

Because Spring.Net.Core also depends on Common.Logging (and they specify 1.2+), I decided to do the same thing.  Common.Logging has since introduced a breaking change to the API in 3.0 so now it is impossible to support both 1.2 and 3.0 of Common.Logging at the same time.

However, this problem only rears its head (supposedly) under the following conditions:

(a) You have an other project / dependency which requires Common.Logging v3.0.0
(b) You are using an old NuGet (older than 2.5). http://docs.nuget.org/create/versioning#dependencies-are-no-longer-unnecessarily-updated-during-package-installation
(c) You used the Update command

Indeed I checked and if you install BoboBrowse.Net into a new MVC 4 project with no other dependencies, it will install Common.Logging 1.2. So, unless one of the other issues is in play here you should not see any breaking behavior.

A possible workaround would be to install the highest supported version of Common.Logging 2.x that works with all of your other dependencies manually *before* installing BoboBrowse.Net.

Since we depend on Spring.Net.Core, the only logical thing to do is the same thing they are doing. In their master branch, they have updated to Common.Logging 3.0.0.

I have managed to do the same, but it [requires that BoboBrowse.Net be built against Common.Logging 3.0.0](https://github.com/net-commons/common-logging/issues/74#issuecomment-69954581). Also, the Common.Logging.NLog references must be removed because they haven't updated that part yet.

I will let you know when this change has been released. Until then, use a version of Common.Logging lower than 3.0.0.

**Comment by dtosato** [comment]
Actually, for my solution, there are two reasons why Common.Logging 3.0.0 is downloaded by NuGet. The first reason is that I have a dependency that requires Common.Logging which is Common.Logging.log4net1210 (as you written). But I can remove this package renouncing to the log coming from bobo.
The second reason is that into the configuration file of NuGet 2.8 (NuGet.Config), I have this setting:
```xml
 <config>
    <add key="dependencyversion" value="Highest" />
 </config>
```
that force NuGet to download the latest version of the packages.

**Comment by dtosato** [comment]
@NightOwl888 
Any update on this problem?

**Comment by NightOwl888** [comment]
I attempted to fix this issue in [3.2.2](https://github.com/NightOwl888/BoboBrowse.Net/releases/tag/v3.2.2).

However, in hindsight I think that the problem is now even worse. The fix I applied was to require 3.0 as a minimum version (which is the same fix that the author of Common.Logging used in Spring.Net). Since Common.Logging 3.0 is not binary compatible with any prior version, it means that if BoboBrowse.Net is installed into an application that has another component that depends on Common.Logging < 3.0, it will break the other component. There are 3 ways out of this scenario:

1. Obtain a new version of the other component that references Common.Logging 3.0+.
2. Rework the Common.Logging/Spring.Net dependencies of BoboBrowse.Net and then compile your own package to work with Common.Logging <3.0.
3. Use a version of BoboBrowse.Net that supports Common.Logging < 3.0 (currently only 3.2.1 does).

I am considering making option 2 easier by putting a compilation argument into the build script and/or creating 2 more NuGet packages that depend on Common.Logging < 3.0 so the new BoboBrowse.Net packages will work with those other 3rd party components. But I am hoping that most/all other component authors that depend on Common.Logging recognize the need to create a package that is available for Common.Logging 3.0+.

It sucks. But the fault is the author of Common.Logging for making a breaking change that is the very definition of DLL hell by making it impossible to support all versions of his component. The [way NuGet is supposed to work](http://blog.davidebbo.com/2011/01/nuget-versioning-part-2-core-algorithm.html) is to choose the lowest common version that works with all dependencies (during a new install anyway, during an upgrade all bets are off). The whole scheme depends on each version being binary backward compatible with the last version, and breaking this chain causes a ripple effect across all packages that depend on it (and packages that depend on those, etc), including BoboBrowse.Net.

That said, it is not a good practice for you to force an update to the highest version because it will prevent the install logic from being able to choose the lowest common version between all of your project's dependencies. The recommendation to NuGet package authors is to keep the top version open-ended on any component *unless there is a known incompatibility* at the time the package is authored. Therefore, when you set the option to `value="Highest"`, you always risk a component breaking as a result of a future unsupported version of a dependency being released.

The reality of NuGet is that they tie the package authors' hands and leave it up to the package consumer to work out incompatibilities between versions. But sound advice is to *always uninstall a package and install the next higher version, never upgrade*. This will ensure that the version constraint logic is respected at the time of the upgrade. See [the discussion here](http://nuget.codeplex.com/discussions/578411#PostContent_1348041) for details.
